src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_Call.cs

```diff
@@ -133,34 +133,137 @@ private void EmbedIfNeedTo(BoundExpression receiver, ImmutableArray<PropertySymb
+
+            method.TryGetThisParameter(out var methodThisParameter);
+            symbolForCompare.TryGetThisParameter(out var interceptorThisParameterForCompare);
+            switch (methodThisParameter, interceptorThisParameterForCompare)
+            {
+                case (not null, null):
+                case (not null, not null) when !methodThisParameter.Type.Equals(interceptorThisParameterForCompare.Type, TypeCompareKind.ObliviousNullableModifierMatchesAny)
+                        || methodThisParameter.RefKind != interceptorThisParameterForCompare.RefKind: // PROTOTYPE(ic): and ref custom modifiers are equal?
+                    this._diagnostics.Add(ErrorCode.ERR_InterceptorMustHaveMatchingThisParameter, interceptsLocationAttributeData.AttributeLocation, methodThisParameter, method);
+                    break;
+                case (null, not null):
+                    this._diagnostics.Add(ErrorCode.ERR_InterceptorMustNotHaveThisParameter, interceptsLocationAttributeData.AttributeLocation, method);
+                    break;
+                default:
+                    break;
+
+            }
+
+            if (needToReduce)
+            {
+                Debug.Assert(!method.IsStatic);
+                arguments = arguments.Insert(0, receiverOpt!);
+                receiverOpt = null;
+
+                var thisParameter = method.ThisParameter;
+
+                if (!argsToParamsOpt.IsDefault)
+                {
+                    var argsToParamsBuilder = ImmutableArray.CreateBuilder<int>(argsToParamsOpt.Length + 1);
+                    argsToParamsBuilder.Add(0); // argument 0 always corresponds to parameter 0 (extension this)
+                    foreach (var paramOrdinal in argsToParamsOpt)
+                    {
+                        // the param that each argument *really* refers to is pushed forward by 1.
+                        // e.g. we're adapting:
+                        // void Type.Method(int param) { }
+                        // void Ext.Method(this Type type, int param) { }
+                        argsToParamsBuilder.Add(paramOrdinal + 1);
+                    }
+
+                    argsToParamsOpt = argsToParamsBuilder.MoveToImmutable();
+                }
+
+                if (!argumentRefKindsOpt.IsDefault)
+                {
+                    argumentRefKindsOpt = argumentRefKindsOpt.Insert(0, thisParameter.RefKind);
+                }
+
+                invokedAsExtensionMethod = true;
+            }
+
+            method = interceptor;
+
+            return (method, receiverOpt, arguments, argsToParamsOpt, argumentRefKindsOpt, invokedAsExtensionMethod);
+        }
+
         public override BoundNode VisitCall(BoundCall node)
         {
             Debug.Assert(node != null);
 
+            var (method, receiverOpt, arguments, argsToParamsOpt, argRefKindsOpt, invokedAsExtensionMethod) = InterceptCallAndAdjustArguments(node);
```

Comment:
> var (method, receiverOpt, arguments, argsToParamsOpt, argRefKindsOpt, invokedAsExtensionMethod) = InterceptCallAndAdjustArguments(node); [](http://example.com/codeflow?start=11&length=137)

in terms of testing, we should think about a test hook that would allow us to attempt to intercept every call across all test scenarios that we currently have. The hook should probably be off by default. We can discuss this in more details offline.

Reply:
Have added a prototype comment for this.